### ISSUE TITLE
C2D messaging ReceiveAsync produces IotHubException (receiverXXX is closed)

### ISSUE THREAD

**Issue by oacsed** [Azure/azure-iot-sdk-csharp#466]
- Deployed on Azure Service Fabric on Windows Server 2016
- Using .NET Core 2.0
- Devices are simulated. We have 100 simulated device and we're doing load tests (say 20 msg/s per device)
- Using Microsoft.Azure.Devices.Client, Version=1.7.1.0

# The issue
We're building some kind of cloud gateway to communicate with devices. For now, we're using simulated devices.
We have a long running Task for each device that is waiting for C2D messages. The code we're using is almost the same as the one provided in the documentation (https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-csharp-csharp-c2d).
Our issue is, `deviceClient.ReceiveAsync();` produces `IotHubException `exceptions with the following stack trace:

```
Microsoft.Azure.Devices.Client.Exceptions.IotHubException:
   at Microsoft.Azure.Devices.Client.DeviceClient+<>c.<ApplyTimeoutMessage>b__75_2 (Microsoft.Azure.Devices.Client, Version=1.7.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Threading.ExecutionContext.Run (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at XXX.YYY.ZZZ+<StartReceiveC2DAsync>d__16.MoveNext (XXX.YYY, Version=1.0.0.0, Culture=neutral, PublicKeyToken=nullXXX.YYY, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null: ..\XXX.YYY, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null: 364)
```
The message of the exception is 

`'receiver1161540' is closed` (where 1161540 is a random number)

# The code
When receiving a message (D2C message) from a device in our cloud gateway, we're checking if we have a cached `DeviceClient `instance. if it's the case, we use that instance to send the message to the IoT Hub. If we don't have such cached instance, we create one using the `CreateDeviceClientAsync `method as shown below. It is responsible to start the C2D listening task, too.

```
private async Task<DeviceClient> CreateDeviceClientAsync(string deviceId, string deviceKey, CancellationToken cancellationToken)
{
    const ushort MaxNumberOfPools = 65535;  // Found by decompiling the Microsoft.Azure.Devices.Client.dll assembly
                                            // We'll need some more thoughts if the number of devices exceeds this number...

    // Only the first call will create the underlying socket
    var auth = new DeviceAuthenticationWithRegistrySymmetricKey(deviceId, deviceKey);
    var client = DeviceClient.Create(
        "XXXX",
        auth,
        new ITransportSettings[]
        {
        new AmqpTransportSettings(TransportType.Amqp_Tcp_Only)
        {
            AmqpConnectionPoolSettings = new AmqpConnectionPoolSettings { Pooling = true, MaxPoolSize = MaxNumberOfPools },
        },
        });

    // Should be cheap for all instance except the first one to use the connection
    await client.OpenAsync();

    this.PrepareC2DMessageReception(deviceId, client, cancellationToken);

    return client;
}

private void PrepareC2DMessageReception(string deviceId, DeviceClient client, CancellationToken cancellationToken)
{
    // Fire and forget style
    this.StartReceiveC2DAsync(deviceId, client, cancellationToken, (id, msg) =>
    {
        SomeDataObject data;
        try
        {
            data = JsonConvert.DeserializeObject<SomeDataObject>(msg);
        }
        catch
        {
            return;
        }

        if (data != null)
        {
            // C2D message received !
            // We will execute some code here (not relevant, so removed)
        }
    });
}

private async void StartReceiveC2DAsync(string deviceId, DeviceClient deviceClient, CancellationToken cancellationToken, Action<string, string> onMessageReceived)
{
    while (true)
    {
        try
        {
            cancellationToken.ThrowIfCancellationRequested();

            // Exception is thrown at the following line:
            var receivedMessage = await deviceClient.ReceiveAsync();
            if (receivedMessage == null)
            {
                continue;
            }

            onMessageReceived?.Invoke(deviceId, Encoding.UTF8.GetString(receivedMessage.GetBytes()));
            await deviceClient.CompleteAsync(receivedMessage);
        }
        catch (Exception ex)
        {
            // Closes the deviceClient in case it wasn't already disposed. Then just returns so it exits the never ending loop
            // The DeviceClient as well as this receive handler will be re-created again
            // next time the device sends some data.
            // Nothing more can be done here
            await this.SafeInvalidateDeviceClient(deviceId, deviceClient, cancellationToken);

            return;
        }
    }
}

private async Task SafeInvalidateDeviceClient(string deviceId, DeviceClient deviceClient, CancellationToken cancellationToken)
{
    try
    {
        // Tries to close the deviceClient
        await deviceClient.CloseAsync();
    }
    catch
    {
        // Silently swallows the exception
    }

    // Now invalidates the cache so next time the device communicates, we'll have a fresh copy
    this.devices.TryRemove(deviceId, out var notUsed);
}
```

# Questions
Is the above code correctly written? We do not especially like the fire and forget task with the infinite loop.
Why is there such exception?

**Comment by RubiaL** [comment]
Can you try the latest version 1.21.1. This issue is possibly fixed there.

**Comment by prmathur-microsoft** [comment]
We believe this issue is fixed in latest release 1.21.1. I'm closing this issue for now. Feel free to reopen if you still see further issues.

**Comment by az-iot-builder-01** [comment]
@oacsed, @RubiaL, @prmathur-microsoft, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
